pkg/controller/usersignup/usersignup_controller.go

```diff
@@ -0,0 +1,195 @@
+	completed := condition.IsConditionTrue(instance.Status.Conditions, toolchainv1alpha1.UserSignupComplete)
+
+	// If the status is not yet provisioning or completed, check for automatic or manual approval
+	if !provisioning && !completed {
+		userApprovalPolicy, err := r.ReadUserApprovalPolicyConfig()
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		// If the signup has been explicitly approved (by an admin), or the user approval policy is set to automatic,
+		// then proceed with the signup
+		if instance.Spec.Approved || userApprovalPolicy == config.UserApprovalPolicyAutomatic {
+			var targetCluster string
+
+			// If a target cluster hasn't been selected, select one from the members
+			if instance.Spec.TargetCluster != "" {
+				targetCluster = instance.Spec.TargetCluster
+			} else {
+				// Automatic cluster selection
+				members := cluster.GetMemberClusters()
+				if len(members) > 0 {
+					targetCluster = members[0].Name
+				}
+    			// TODO how do we handle the situation where no members are available?
+
+			}
+
+			// Provision the MasterUserRecord
+			userAccounts := []toolchainv1alpha1.UserAccountEmbedded{
+				{
+					TargetCluster: targetCluster,
+				},
+			}
+
+			mur := &toolchainv1alpha1.MasterUserRecord{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: instance.Spec.UserID,
+					Namespace: config.GetOperatorNamespace(),
+				},
+				Spec: toolchainv1alpha1.MasterUserRecordSpec{
+					UserID: instance.Spec.UserID,
+					UserAccounts: userAccounts,
+				},
+			}
+
+			err := r.client.Create(context.TODO(), mur)
+			if err != nil {
+				// Error creating the MasterUserRecord - requeue the request.
+				reqLogger.Error(err, "Error creating MasterUserRecord", "UserID", instance.Spec.UserID, "TargetCluster", targetCluster)
+				return reconcile.Result{}, err
+			}
+			reqLogger.Info("Created MasterUserRecord", "UserID", instance.Spec.UserID, "TargetCluster", targetCluster)
+
+			// Update the UserSignup status conditions: set PendingApproval to false, Provisioning to true
+			updatedConditions, updated := commonCondition.AddOrUpdateStatusConditions(instance.Status.Conditions, toolchainv1alpha1.Condition{
+				Type: toolchainv1alpha1.UserSignupPendingApproval,
+				Message: "Approved",
+				Status: corev1.ConditionFalse,
+				Reason: "",
+			},
+				toolchainv1alpha1.Condition{
+					Type: toolchainv1alpha1.UserSignupProvisioning,
+					Message: "Provisioning",
+					Status: corev1.ConditionTrue,
+					Reason: "",
+				})
+
+			// If the condition values were updated, post the changes
+			if updated {
+				instance.Status.Conditions = updatedConditions
+				err = r.client.Update(context.TODO(), instance)
+				if err != nil {
+					reqLogger.Error(err, "Error updating UserSignup Status conditions", "UserID", instance.Spec.UserID)
+					// Error updating the status - requeue the request.
+					return reconcile.Result{}, err
```

Comment:
do we have a test to verify that if an error occurred here and the request is requeued, we won't end up with 2 `MasterUserRecord` CRs for the same User? I believe that https://github.com/codeready-toolchain/host-operator/pull/27/files#diff-5b4b18a0fc2e0d708e4e6334cdb6bc69R147 will created a new one

Reply:
We can't have two MasterUserRecord instances because they would have a name conflict.  I guess this is one of the situations where the `r.client.Create()` function above would return an error.

Reply:
but in this case, you could check `errors.IsAlreadyExists(err)` and if it is `true` then just log the error and return `reconcile.Result{}, nil`

Reply:
I've added a test for this.  At the moment though, I'm still returning an error from the reconciler if the MUR already exists.  Do we really want the reconciliation to die here if the MUR already exists? @MatousJobanek what do you think the workflow would be for an admin user in the case of this scenario?

Reply:
> At the moment though, I'm still returning an error from the reconciler if the MUR already exists. 

it's fine, just maybe make the delay of the next reconciler a bit longer.

> @MatousJobanek what do you think the workflow would be for an admin user in the case of this scenario?

There are two scenarios based on the cause of the conflict:
1. delete the existing MUR (in case it's kind of old zombie)
2. manually set the status of UserSignup to complete (assuming that the MUR is the intended one)

Reply:
I've set the requeue delay to 1 hour, if you think that's too long please let me know.  My consideration was of the first scenario that you described - the UserSignup fails because of an existing MUR, the admin user deletes the MUR and the reconcile will then execute again successfully (after a delay).